Fix websocket when reading small amount of data

jukkar opened the pull request:
Also add unit test for that.

Fixes #21989

jukkar committed:
net: websocket: Fix receiving websocket data

Make sure that if we receive websocket data in small chunks,
the parsing and returning of data to caller is done properly.

Fixes #21989

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: websocket: Add RX unit tests for websocket API

Add simple tests for testing receiving websocket data.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon commented on subsys/net/lib/websocket/websocket.c:
```diff
@@ -443,12 +443,25 @@ static int websocket_ioctl_vmeth(void *obj, unsigned int request, va_list args)
+	if (mask_shift < 0) {
+		mask_shift_ptr = &zero;
+	} else {
+		mask_shift_ptr = &mask_shift;
```
How var of type `int*` can be assigned with address of `s8_t`?

jukkar replied:
Ah, simple typo error. I will refactor this function or probably remove it altogether as it is now looking quite convoluted.

pfalcon commented on subsys/net/lib/websocket/websocket.c:
```diff
@@ -443,12 +443,25 @@ static int websocket_ioctl_vmeth(void *obj, unsigned int request, va_list args)
+
+	/* Special handling when receiving data. As we cannot guarantee that
+	 * read data is received in four byte blocks, we need to pick proper
+	 * byte from the shift mask. When sending data, the mask_shift is set
```
> When sending data, the mask_shift is set to -1

Why do we call this routine at all "when sending data"?

pfalcon replied:
Ah, nevermind, it's the structure of the operations here got me confused here. I just remember that masking of data happens in one direction, but not another (e.g. server->client, but not vice-versa, I don't remember which exactly). So, the question then is why we need this special -1 value, why don't just pass 0 as an arg.

jukkar replied:
When unmasking data, we need to select correct byte from 32-bit mask value. This is an issue if we are reading less than 4 bytes. Anyway, I will refactor this this part in order to avoid these magic values.

pfalcon commented on subsys/net/lib/websocket/websocket.c:
```diff
@@ -443,12 +443,25 @@ static int websocket_ioctl_vmeth(void *obj, unsigned int request, va_list args)
 
 static void websocket_mask_payload(u8_t *payload, size_t payload_len,
-				   u32_t masking_value)
+				   u32_t masking_value, s8_t mask_shift)
```
Are you sure this argument should be of s8_t type, instead of int? Some compilers may produce less efficient code in thus case (effectively, adding sign extension cast from s8_t to full register size where all operations are handled then).

pfalcon reviewed (commented):
Some type safety issues are evident.

pfalcon commented on subsys/net/lib/websocket/websocket.c:
```diff
@@ -643,10 +656,24 @@ int websocket_recv_msg(int ws_sock, u8_t *buf, size_t buf_len,
 	int ret;
 
+#if defined(CONFIG_NET_TEST)
+	/* Websocket unit test does not use socket layer but feeds
```
I would image we'd end up with such "test handler" in the code eventually, e.g. for fault injection. But is it warranted here now? What tests you can't perform via socket layer which required such contraptions? In the end, users will user this API via socket layer, and that's what we're expected to test.

jukkar replied:
I do not want to implement websocket server just to test this receive function. Using sockets would be nice of course but here sockets just complicate things unnecessarily. It should not really matter where the input data comes from to this function when testing it.

jukkar committed:
tests: net: websocket: Add TX unit tests for websocket API

Add simple tests for testing sent and received websocket data.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: websocket: Verify the received data properly

Make sure that we do not calculate terminating \n when comparing
the received data to sent data because the \n is not part of
the lorem_ipsum buffer.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
New version that adds TX unit test and has a complete rewrite of the `websocket_recv_msg()` function. The recv function was too convoluted and not handling all the corner cases. While testing I also noticed that the sample app was doing things slightly wrong and it missed some error cases.

jukkar commented:
ping @nashif Any idea why sanity fails constantly in shippable? When running manually, I see no errors (tried multiple times):
```
$ sanitycheck -s tests/net/socket/websocket/net.socket.websocket -p native_posix_64 -c
Install the anytree module to use the --test-tree option
Install tabulate python module with pip to use --device-testing option.
Deleting output directory zephyr/sanity-out
INFO    - JOBS: 8
INFO    - Building initial testcase list...
INFO    - 1 test configurations selected, 730 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Total complete:    1/   1  100%  skipped:    0, failed:    0
INFO    - 1 of 1 tests passed (100.00%), 0 failed, 0 skipped with 0 warnings in 5.19 seconds
INFO    - In total 742 test cases were executed on 1 out of total 220 platforms (0.45%)
```

jukkar commented:
No functionality changes, just trying to tweak config options in order to get sanity to pass.
